Interpolation without Interpolator ParallelComponent for time-dependent maps.

markscheel opened the pull request:
## Proposed changes

Previously, interpolation without an Interpolator ParallelComponent didn't work correctly if you had an InterpolationTarget that had points that were time-independent in any frame other than the Grid frame, and if you had a time-dependent map.  In particular, this PR does the following:
 - Previously, for interpolation without the Interpolator ParallelComponent, each Element stored the block_logical_coordinates of the interpolation points, which were computed in the registration phase; but block_logical_coordinates should be time-dependent in some cases, so that behavior was wrong.  So now each element stores the interpolation points in whatever frame the InterpolationTarget designates; the points must be time-independent in that frame, and those points are computed and stored during the Registration phase as before.  
 - The block_logical_coordinates are now computed in intrp::Events::InterpolateWithoutInterpComponent, which is able to pass the correct temporal_id and the correct FunctionsOfTime into that calculation.
 - Previously, some functions in InterpolationTarget_detail internally computed either the original coordinates of the interpolation points or the block_logical_coordinates. These functions were modified to be consistent with the new behavior.
 - InterpolationTargetVarsFromElement needs the block_logical_coordinates of the interpolation points (but uses only part of block_logical_coordinates: the size information and whether there are any points that are not in any blocks); Previously InterpolationTargetVarsFromElement was computing block_logical_coordnates internally (and inconsistently).  Now InterpolationTargetVarsFromElement takes block_logical_coordinates as a function argument, so the correct block_logical_coordinates can be passed in.
 - Tests were modified to be consistent with the new behavior.

Note that this PR conflicts with parts of #4022 and should replace those parts.  Other parts of #4022 (in particular some of the static_asserts for non-sequential InterpolationTargets and time-dependent maps) should be fine and might even be incorporated into this PR but I have not done that yet.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
No changes needed for the user, but anyone changing the interpolator code will see differences:
- intrp::Vars::PointInfoTag now has a different type.
- InterpolationTargetVarsFromElement and some of the functions in InterpolationTarget_detail now take different arguments.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel committed:
maps_are_time_dependent takes GlobalCache, not DataBox.

Sometimes it is not clear which DataBox to pass, if you don't have
one handy or you are not already in a ParallelComponent.
But you always have a GlobalCache handy.

markscheel committed:
InterpolationTargetVarsFromElement: pass in block_logical_coords

Previously it computed block_logical_coords internally,
but block_logical_coords is more complicated to compute nowadays,
and InterpolationTargetVarsFromElement doesn't have all the info
it needs to compute block_logical_coords.

markscheel committed:
Test_InterpolateDuringSelfStart: use a consistent frame.

Frame needs to match the one in InterpolateOnElementTestHelpers.

markscheel committed:
intrp_detail::block_logical_coords takes coords as input.

So now you can pass in the coords in the frame that you want the output.

The old overloads of intrp_detail::block_logical_coords are still
there and are still used; those compute the coords internally
by calling the InterpolationTarget.

The new overload is used for the case without an Interpolator
ParallelComponent, where the coords are stored.

markscheel committed:
PointInfoTag holds coords in InterpolationTarget's frame.

Previously it held block_logical_coords, but those coords can't
always be computed ahead of time.

Also:

- InterpolationTargetSendPoints now sends InterpolationTarget's coords.
- ElementReceiveInterpPoints now receives InterpolationTarget's coords.
- InterpolateWithoutInterpComponent now uses the InterpolationTarget
  coords that it has stored, and computes block_logical_coords when
  it is called, so that it can pass the correct temporal_id into
  the block_logical_coords call.

These changes are to allow the interpolator to work correctly
when interpolating without the Interpolator parallel component,
when interpolating onto points that are time-independent in
the frame of the InterpolationTarget.

knelli2 commented on src/ParallelAlgorithms/Interpolation/InterpolationTargetDetail.hpp:
```diff
@@ -522,13 +521,31 @@ auto block_logical_coords(const db::DataBox<DbTags>& box,
+  return ::block_logical_coordinates(domain, coords);
+}
+
+/// Version of block_logical_coords that computes the interpolation
+/// points on the fly.  This version is called when the interpolation
+/// points are (or might be) time-dependent in the
+/// InterpolationTargetTag's frame.
```
I think it would be helpful to add to this comment that this is primarily used without an Interpolator parallel component. (Unless that's the other overload? I think some clarification on which is used when you have the Interpolator parallel component would be nice)

knelli2 reviewed (changes requested):
Looks good. I think #4022 can be replaced in favor of this PR. If there's anything you want to take from that PR (like sanity checks, etc...) you can just add them here.

Squash any changes directly.

markscheel committed:
Update Docs in InterpolationTarget.hpp

markscheel commented:
Made changes and squashed.  I added a new commit because I realized that the docs in InterpolationTarget.hpp were out of date (and some of those docs indicated that a future change would be essentially this PR), and I incorporated a static_assert and some docs changes from #4022 .   I don't think the changes in #4022 that add time-dependence to some tests are relevant anymore; one of those tests is now time-independent and for the other the time-dependence isn't really tested in a meaningful way.

nilsdeppe commented on src/ParallelAlgorithms/Interpolation/InterpolationTargetDetail.hpp:
```diff
@@ -522,13 +521,34 @@ auto block_logical_coords(const db::DataBox<DbTags>& box,
+template <typename InterpolationTargetTag, typename DbTags,
+          typename Metavariables, typename TemporalId>
+auto block_logical_coords(const db::DataBox<DbTags>& box,
+                          Parallel::GlobalCache<Metavariables>& cache,
```
I'd expect the `cache` can be constant. It only needs to be non-const if being used for communication

nilsdeppe commented on src/ParallelAlgorithms/Interpolation/InterpolationTargetDetail.hpp:
```diff
@@ -468,27 +468,28 @@ void add_received_variables(
+template <typename InterpolationTargetTag, typename Metavariables,
+          typename TemporalId>
+auto block_logical_coords(
+    Parallel::GlobalCache<Metavariables>& cache,
```
I'd expect the `cache` can be constant. It only needs to be non-const if being used for communication

nilsdeppe commented on src/ParallelAlgorithms/Interpolation/Actions/ElementReceiveInterpPoints.hpp:
```diff
@@ -37,19 +37,17 @@ struct ElementReceiveInterpPoints {
   static void apply(
       db::DataBox<DbTags>& box, Parallel::GlobalCache<Metavariables>& /*cache*/,
```
`const & cache`?

nilsdeppe reviewed (changes requested):
LGTM, just a few small things. Squash immediately. For future PRs, some of the short commit names should be more verby. E.g. `Add X` `Change X to Y` etc.

markscheel commented:
Made some caches const, squashed rebased.
